### PR TITLE
[BUG] maybe_upcast_putmast also handle ndarray

### LINKED ISSUE #23823: BUG/Internals: maybe_upcast_putmask
In the context of #23192 (and #23604 / #23606), I want to use `pandas.core.dtypes.cast.maybe_upcast_putmask`,  because it solves exactly the problem I need it to solve.

Unfortunately, it does not work as advertised (and I already found the culprit).
The docstring says: 
```
def maybe_upcast_putmask(result, mask, other):
    """
    A safe version of putmask that potentially upcasts the result

    Parameters
    ----------
    result : ndarray
        The destination array. This will be mutated in-place if no upcasting is
        necessary.
    mask : boolean ndarray
    other : ndarray or scalar
        The source array or value
```
in other words, it expects `result` and `other` to be ndarrays. Curiously enough, in some branches, it only works for `Series` and produces wrong results for `ndarray`, e.g.

```
>>> import pandas as pd
>>> import numpy as np
>>> s = pd.Series([10, 11, 12])
>>> t = pd.Series([np.nan, 61, np.nan])
>>> from pandas.core.dtypes.cast import maybe_upcast_putmask
>>> result, _ = maybe_upcast_putmask(s, np.array([False, True, False]), t)
>>> result  # correct
0    10
1    61
2    12
dtype: int64
>>> result, _ = maybe_upcast_putmask(s.values, np.array([False, True, False]), t.values)
>>> result  # incorrect
array([10., nan, 12.])
```

This is because the code does 
```
try:
    [...]
    new_result = result.values.copy()
    [...]
    return [...]
except: 
    # do something else
```

which actually expects a `Series` (since `.values` won't ever work on  an ndarray).

### PR DESCRIPTION
**Pull request by makbigc**
-  closes #23823
-  2 tests added
-  passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
-  whatsnew entry

Follow h-vetinari's footstep. `maybe_upcast_putmask` was left untouched in #25425. Try to fix the bug so that `maybe_upcast_putmask` can also handle `ndarray` as well as `Series`.

### CODE REVIEW

File: pandas/tests/dtypes/cast/test_upcast.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def test_upcast_np_array():
+    result = np.array([10, 11, 12])
+    mask = np.array([False, True, False])
+    other = np.array([np.nan, 61, np.nan])
+    expected = np.array([10, 61, 12])
+
+    result, _ = maybe_upcast_putmask(result, mask, other)
+    tm.assert_numpy_array_equal(result, expected)
```
Comment (gfyoung) [inline pandas/tests/dtypes/cast/test_upcast.py]: I suspect these tests already have a home in another, existing test file.

Comment (gfyoung) [reply pandas/tests/dtypes/cast/test_upcast.py]: Also, reference the issue number as a comment under each test definition.

File: pandas/tests/dtypes/cast/test_upcast.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import numpy as np
+
+from pandas.core.dtypes.cast import maybe_upcast_putmask
+
+from pandas import Series
+from pandas.util import testing as tm
+
+
+def test_upcast_series():
```
Comment (jreback) [inline pandas/tests/dtypes/cast/test_upcast.py]: can you parameterize these (both tests) as I suspect a myriad of cases (just a couple for now, e.g. int -> float you can start with)

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -209,7 +209,10 @@ def changeit():
                 om = other[mask]
                 om_at = om.astype(result.dtype)
                 if (om == om_at).all():
-                    new_result = result.values.copy()
+                    if isinstance(result, np.ndarray):
+                        new_result = result.copy()
+                    else:
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: what other types are here? a Series?

this should be converted at a higher level (at the beginning of the function)

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -105,7 +105,7 @@ Performance Improvements
 Bug Fixes
 ~~~~~~~~~
 
--
+- :func:`maybe_upcast_putmask` does not replace the element of the ``ndarray`` correctly (:issue:`23823`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this doesn't warrant a whatsnew as its an internal impl detail

**Review by jreback — CHANGES_REQUESTED** [review]
what problem is this solving?

**Comment by makbigc** [comment]
I investigated further and found that I misunderstood the usage of `maybe_upcast_putmask`. 

The result array should be replaced with the element of other array starting from the beginning, not in the position of mask. This behavior is due to np.place

 `changeit` isn't called.
```
In [4]: arr1 = np.array([1,2,3])                                                

In [5]: other = np.array([61,62,63])                                            

In [6]: mask = np.array([False, True, False])                                   

In [7]: result, _ = maybe_upcast_putmask(arr1, mask, other)                     

In [8]: result                                                                  
Out[8]: array([ 1, 61,  3])
```

`changeit` is called.
```
In [24]: arr2 = np.array([1,2,3])                                                                                                           

In [25]: other2 = np.array([61.1, 62.2, 63.3])                                                                                              

In [26]: mask = np.array([False, True, False])                                                                                              

In [27]: result, _ = maybe_upcast_putmask(arr2, mask, other2)                                                                               

In [28]: result                                                                                                                             
Out[28]: array([ 1. , 61.1,  3. ])
```

`changeit` isn't called.
```
In [15]: arr2 = np.arange('2019-01-01', '2019-01-04', dtype='datetime64[D]')    

In [16]: other2 = np.arange('2018-01-01', '2018-01-04', dtype='datetime64[D]')  

In [17]: mask = np.array([False, True, False])                                  

In [18]: result, _ = maybe_upcast_putmask(arr2, mask, other2)                   

In [19]: result                                                                 
Out[19]: array(['2019-01-01', '2018-01-01', '2019-01-03'], dtype='datetime64[D]')
```
So the behavior for Series as an input is not correct. And `maybe_upcast_putmask` does not expect a series input.

`changeit` is called.
```
>>> s = pd.Series([10, 11, 12])
>>> t = pd.Series([np.nan, 61, np.nan])
>>> from pandas.core.dtypes.cast import maybe_upcast_putmask
>>> result, _ = maybe_upcast_putmask(s, np.array([False, True, False]), t)
>>> result  # correct
0    10
1    61
2    12
dtype: int64
```
Examples can be added into the function docstring to elucidate the usage. Should I do it in this PR?

File: doc/source/whatsnew/v0.25.0.rst (hunk 4 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: can you remove

**Review by jreback — CHANGES_REQUESTED** [review]
can you update with your new tests. also you can do an assert on the input of the function itself (e.g. assert its an ndarray)

**Comment by makbigc** [comment]
When adding the tests, the following bug is found and fixed. 

```
In [4]: arr = np.arange(10, 15)                                                 

In [5]: other = np.array([61, 62])                                              

In [6]: mask = np.array([False, True, False, True, True])                       

In [7]: result, changed = maybe_upcast_putmask(arr, mask, other)                
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-7-3ec9bc405a70> in <module>
----> 1 result, changed = maybe_upcast_putmask(arr, mask, other)

~/Code/pandas/pandas/core/dtypes/cast.py in maybe_upcast_putmask(result, mask, other)
    242             else:
    243 
--> 244                 if isna(other[mask]).any():
    245                     return changeit()
    246 

IndexError: boolean index did not match indexed array along dimension 0; dimension is 2 but corresponding boolean dimension is 5
```

File: pandas/tests/dtypes/cast/test_upcast.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import numpy as np
+import pytest
+
+from pandas.core.dtypes.cast import maybe_upcast_putmask
+
+from pandas import Series
+from pandas.util import testing as tm
+
+
+def test_upcast_error():
```
Comment (jreback) [inline pandas/tests/dtypes/cast/test_upcast.py]: are there more cases to consider? e.g. non-ndarray

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments, merge master, ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @makbigc keep em coming!

**Pull request merged**